src/Tools/ExternalAccess/LiveShare/Tagger/SyntacticClassificationTaggerProvider.TagComputer.cs

```diff
@@ -0,0 +1,256 @@
+
+                using var tagSpans = SharedPools.Default<List<ITagSpan<IClassificationTag>>>().GetPooledObject();
+                ClassificationUtilities.Convert(_typeMap, snapshot, classifiedSpans, tagSpans.Object.Add);
+                ClassificationUtilities.ReturnClassifiedSpanList(classifiedSpans);
+
+                var tagList = new TagSpanList<IClassificationTag>(tagSpans.Object);
+                Interlocked.Exchange(ref this._lastProcessedTagList, tagList);
+
+                this.TagsChanged?.Invoke(this, new SnapshotSpanEventArgs(snapshot.GetFullSpan()));
+            }
+
+            public event EventHandler<SnapshotSpanEventArgs> TagsChanged;
+
+            public IEnumerable<ITagSpan<IClassificationTag>> GetTags(NormalizedSnapshotSpanCollection spans)
+            {
+                // A background thread might be updating lastProcessedTagList.
+                var tagList = Volatile.Read(ref this._lastProcessedTagList);
+                return tagList?.GetTags(spans) ?? EmptyTagSpanEnumerable;
+            }
+
+            private void OnDocumentActiveContextChanged(object sender, DocumentActiveContextChangedEventArgs args)
+            {
+                var workspace = this._workspace;
+                if (workspace != null && workspace == args.Solution.Workspace)
+                {
+                    EnqueueProcessSnapshotAsync(args.NewActiveContextDocumentId);
+                }
+            }
+
+            private void OnDocumentOpened(object sender, DocumentEventArgs args)
+            {
+                EnqueueProcessSnapshotAsync(args.Document.Id);
+            }
+
+            private void OnWorkspaceChanged(object sender, WorkspaceChangeEventArgs args)
+            {
+                // We're getting an event for a workspace we already disconnected from
+                var workspace = this._workspace;
+                if (args.NewSolution.Workspace != workspace)
+                {
+                    // we are async so we are getting events from previous workspace we were associated with
+                    // just ignore them
+                    return;
+                }
+
+                switch (args.Kind)
+                {
+                    case WorkspaceChangeKind.ProjectChanged:
+                        {
+                            var documentId = workspace.GetDocumentIdInCurrentContext(this._textBuffer.AsTextContainer());
+                            if (documentId?.ProjectId == args.ProjectId)
+                            {
+                                EnqueueProcessSnapshotAsync();
+                            }
+                            break;
+                        }
+                    case WorkspaceChangeKind.DocumentChanged:
+                        {
+                            EnqueueProcessSnapshotAsync(args.DocumentId);
+                            break;
+                        }
+                }
+            }
+        }
+    }
+}
```

Comment:
this feels like a huge amount of complexity that we wouldn't want to have.

Reply:
well. people implemented this way since they didn't want the huge complexity Roslyn tagger has.